### DISCUSSION TITLE
HTTPS Nginx Reverse Proxy with Code-Server Docker Container

### DISCUSSION THREAD

**Discussion by DiefBell** [coder/code-server#3646] [Q&A]
Hi,

I'm running both the Nginx and Code-Server Docker images and I'm trying to get nginx to act as a reverse proxy for proxying code.jarvis-test to the code-server container. I've got it working fine with http, and I can get https working to show just a standard webpage. I can directly access code-server on https if I bypass the reverse-proxy/

I've played around with the PROXY_DOMAIN environment variable (though I'm not really sure what it does), settings the code-server container to http and https, and using various headers I found documented online.

Below is my nginx.conf
```
events {
	worker_connections 1024;
}

http {

	# ============ HTTP redirect to HTTPS ============
	server {
		listen  80 default_server;
		listen [::]:80 default_server;
		server_name *.jarvis-test;

		location / {
			return 301 https://$host$request_uri;
		}
	}
	
	# ============ ROOT ============
	server {
		listen 443 ssl;
		server_name jarvis-test;

		ssl_certificate          /etc/nginx/certs/jarvis.crt;
		ssl_certificate_key   /etc/nginx/certs/jarvis.key;

		root     /var/www/html;
		index   index.html index.htm;
	}
 
	# ============ CODE-SERVER ============
	server {
		listen 443 ssl;
		server_name code.jarvis-test;

		ssl_certificate		       /etc/nginx/certs/jarvis.crt;
		ssl_certificate_key 	       /etc/nginx/certs/jarvis.key;

		location / {
			# Reverse proxy to code-server Docker container
			proxy_pass  		https://code-server:8443/;
			
			proxy_set_header 	Host $host;
			proxy_set_header 	Upgrade $http_upgrade;
			proxy_set_header 	Connection upgrade;
			proxy_set_header 	Accept-Encoding gzip;
		}
	}

}
```

Any advice on my setup would be greatly appreciated!

**Comment by jsjoeio** [comment]
Not super familiar with Nginx but we have some docs [here](https://github.com/cdr/code-server/blob/main/docs/guide.md#nginx) and you can read about some environment variables related to proxies [here](https://github.com/cdr/code-server/blob/main/docs/FAQ.md#http_proxy-https_proxy-no_proxy). Let us know if that helps!

**Reply by DiefBell** [reply]
Unfortunately I've followed the docs but they're not working out for me.

Below is the error (502 Bad Gateway) from the nginx container in case anyone is able to help:
```
nginx          | 2021/06/22 16:45:24 [error] 24#24: *8 connect() failed (111: Connection refused) while connecting to upstream,
client: 192.168.1.3,
server: code.jarvis-test,
request: "GET /static/387b12ef4ca404ffd39d84834e1f0776e9e3c005/usr/local/share/.config/yarn/global/node_modules/code-server/dist/serviceWorker.js HTTP/1.1",
upstream: "https://172.19.0.2:443/static/387b12ef4ca404ffd39d84834e1f0776e9e3c005/usr/local/share/.config/yarn/global/node_modules/code-server/dist/serviceWorker.js",
host: "code.jarvis-test",
referrer: "https://code.jarvis-test/static/387b12ef4ca404ffd39d84834e1f0776e9e3c005/usr/local/share/.config/yarn/global/node_modules/code-server/dist/serviceWorker.js"
```

Really not sure if this is an issue with my nginx setup or my code-server setup.

**Reply by jsjoeio** [reply]
Hmm... I wish had better tips for debugging your setup. Maybe @code-asher can help

**Reply by code-asher** [reply]
I think connection refused usually means the upstream isn't running (i.e. code-server isn't running) or it can't connect for some network-related reason.

So my thinking is maybe the NGINX Docker container isn't able to access the code-server Docker container? Or isn't able to access code-server inside that container, maybe something to do with the ports.

If you have the configs or commands for how you are spinning up the containers we might be able to figure something out from that.

**Comment by DiefBell** [comment]
The container is definitely all working because I can directly access it by HTTPS, and NGINX can proxy to it fine in HTTP. Think I'll ask more generally on Stack Overflow since I think this is more Nginx than Vscode
